File: computing/interview/design-patterns/structural/flyweight/python/flyweight_demo.py
from __future__ import annotations


class Glyph:
    def __init__(self, symbol: str) -> None:
        self._symbol = symbol

    def render(self, font_size: int, x: int, y: int) -> None:
        print(f"Rendering '{self._symbol}' size={font_size} at ({x}, {y})")


class GlyphFactory:
    def __init__(self) -> None:
        self._cache: dict[str, Glyph] = {}

    def get(self, symbol: str) -> Glyph:
        try:
            return self._cache[symbol]
        except KeyError:
            glyph = Glyph(symbol)
            self._cache[symbol] = glyph
            return glyph


def main() -> None:
    factory = GlyphFactory()
    text = "FLY"
    x = 0
    for ch in text:
        factory.get(ch).render(12, x, 0)
        x += 10


if __name__ == "__main__":
    main()

File: computing/interview/design-patterns/structural/flyweight/csharp/FlyweightDemo.cs
using System;
using System.Collections.Generic;

namespace Interview.DesignPatterns.Structural.Flyweight
{
    public sealed class Glyph
    {
        private readonly char _symbol;

        public Glyph(char symbol)
        {
            _symbol = symbol;
        }

        public void Render(int fontSize, int x, int y)
        {
            Console.WriteLine($"Rendering '{_symbol}' size={fontSize} at ({x}, {y})");
        }
    }

    public sealed class GlyphFactory
    {
        private readonly Dictionary<char, Glyph> _cache = new();

        public Glyph Get(char symbol)
        {
            if (_cache.TryGetValue(symbol, out var glyph))
            {
                return glyph;
            }

            glyph = new Glyph(symbol);
            _cache[symbol] = glyph;
            return glyph;
        }
    }

    public static class Demo
    {
        public static void Main()
        {
            var factory = new GlyphFactory();
            const string text = "FLY";
            var x = 0;
            foreach (var ch in text)
            {
                factory.Get(ch).Render(12, x, 0);
                x += 10;
            }
        }
    }
}
